lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ContextManager.cs

```diff
@@ -0,0 +1,147 @@
+        /// </summary>
+        /// <param name="totalContexts"></param>
+        internal ContextManager(int totalContexts)
+        {
+            _totalExpectedContexts = totalContexts;
+        }
+
+        /// <summary>
+        /// Returns the collection of IActiveContext
+        /// </summary>
+        internal ICollection<IActiveContext> ActiveContexts
+        {
+            get { return _activeContexts.Values; }
+        }
+
+        /// <summary>
+        /// Returns the difference between the number of expected IActiveContext and actually number of IActiveContext. 
+        /// </summary>
+        internal int NumberOfMissingContexts
+        {
+            get { return _totalExpectedContexts - _activeContexts.Count; }
+        }
+
+        /// <summary>
+        /// Checks if all the requested contexts are received. 
+        /// </summary>
+        internal bool AllContextsReceived
+        {
+            get { return NumberOfMissingContexts == 0; }
+        }
+
+        /// <summary>
+        /// Adding an IActiveContext to the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext already exists.
+        /// </summary>
+        /// <param name="activeContext"></param>
+        internal void Add(IActiveContext activeContext)
+        {
+            if (_activeContexts.ContainsKey(activeContext.Id))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] received already exists.", activeContext.Id);
+                Exceptions.Throw(new ApplicationException(msg), Logger);
+            }
+            _activeContexts.Add(activeContext.Id, activeContext);
+        }
+
+        /// <summary>
+        /// Remove an IActiveContext from the ActiveContext collection
+        /// Throw ApplicationException if the IActiveContext doesn't exist.
+        /// </summary>
+        /// <param name="activeContextId"></param>
+        internal void Remove(string activeContextId)
+        {
+            if (!_activeContexts.ContainsKey(activeContextId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The context [{0}] to be removed does not exist.", activeContextId);
+                Exceptions.Throw(new ApplicationException(msg), Logger);
+            }
+            _activeContexts.Remove(activeContextId);
+        }
+
+        /// <summary>
+        /// Returns the current number of IActiveContext in the ActiveContext collection
+        /// </summary>
+        internal int NumberOfActiveContext
+        {
+            get { return _activeContexts.Count; }
+        }
+
+        /// <summary>
+        /// Given an IFailedEvaluator, remove associated IActiveContext from the collection
+        /// Throw SystemException if associated IActiveContext doesn't exist.
+        /// Throw IMRUSystemException if more than one IActiveContext are associated with the IFailedEvaluator
```

Comment:
I'd add "Current IMRU driver assumes that "there is only one context associated with the IFailedEvalutor" as an explanation of this exception (we have this comment in test but not in the main code)